File: app/src/main/java/yancey/chelper/android/common/dialog/InputStringDialog.java
/**
 * It is part of CHelper. CHelper is a command helper for Minecraft Bedrock Edition.
 * Copyright (C) 2025  Yancey
 * <p>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * <p>
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * <p>
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package yancey.chelper.android.common.dialog;

import android.content.Context;
import android.os.Bundle;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Objects;
import java.util.function.Consumer;

import yancey.chelper.R;

/**
 * 输入文字的对话框
 */
public class InputStringDialog extends FixedDialog {

    // 标题，取消按钮文本，确认按钮文本，默认文本
    @Nullable
    private String title, confirm, cancel, defaultInput;
    // 取消按钮按下后的事件
    @Nullable
    private Consumer<String> onCancel;
    @Nullable
    private Consumer<String> onConfirm;

    public InputStringDialog(@NonNull Context context) {
        super(context);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.dialog_input_string);
        TextView tv_title = findViewById(R.id.title);
        EditText et_title = findViewById(R.id.input);
        TextView btn_confirm = findViewById(R.id.confirm);
        TextView btn_cancel = findViewById(R.id.cancel);
        Objects.requireNonNull(tv_title);
        Objects.requireNonNull(et_title);
        Objects.requireNonNull(btn_confirm);
        Objects.requireNonNull(btn_cancel);
        tv_title.setText(Objects.requireNonNullElse(title, getContext().getString(R.string.title)));
        et_title.setText(defaultInput);
        btn_confirm.setText(Objects.requireNonNullElse(confirm, getContext().getString(R.string.confirm)));
        btn_cancel.setText(Objects.requireNonNullElse(cancel, getContext().getString(R.string.cancel)));
        btn_confirm.setOnClickListener(view -> {
            if (onConfirm != null) {
                onConfirm.accept(et_title.getText().toString());
            }
            dismiss();
        });
        btn_cancel.setOnClickListener(view -> {
            if (onCancel != null) {
                onCancel.accept(et_title.getText().toString());
            }
            dismiss();
        });
    }

    public InputStringDialog title(String title) {
        this.title = title;
        return this;
    }

    public InputStringDialog defaultInput(String defaultInput) {
        this.defaultInput = defaultInput;
        return this;
    }

    public InputStringDialog onConfirm(String confirm) {
        this.confirm = confirm;
        return this;
    }

    public InputStringDialog onConfirm(Consumer<String> onConfirm) {
        this.onConfirm = onConfirm;
        return this;
    }

    @SuppressWarnings("unused")
    public InputStringDialog onConfirm(String confirm, Consumer<String> onConfirm) {
        return onConfirm(confirm).onConfirm(onConfirm);
    }

    public InputStringDialog onCancel(String cancel) {
        this.cancel = cancel;
        return this;
    }

    public InputStringDialog onCancel(Consumer<String> onCancel) {
        this.onCancel = onCancel;
        return this;
    }

    @SuppressWarnings("unused")
    public InputStringDialog onCancel(String cancel, Consumer<String> onCancel) {
        return onCancel(cancel).onCancel(onCancel);
    }

}

File: app/src/main/java/yancey/chelper/ui/common/dialog/InputStringDialog.kt
/**
 * It is part of CHelper. CHelper is a command helper for Minecraft Bedrock Edition.
 * Copyright (C) 2025  Yancey
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package yancey.chelper.ui.common.dialog

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.input.TextFieldState
import androidx.compose.foundation.text.input.rememberTextFieldState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import yancey.chelper.R
import yancey.chelper.ui.common.CHelperTheme
import yancey.chelper.ui.common.widget.Divider
import yancey.chelper.ui.common.widget.DividerVertical
import yancey.chelper.ui.common.widget.Text
import yancey.chelper.ui.common.widget.TextField

@Composable
fun InputStringDialog(
    onDismissRequest: () -> Unit,
    title: String = stringResource(R.string.dialog_input_string_title),
    textFieldState: TextFieldState,
    onConfirm: () -> Unit
) {
    Dialog(onDismissRequest = onDismissRequest) {
        Column(
            modifier = Modifier
                .clip(RoundedCornerShape(10.dp))
                .background(CHelperTheme.colors.background)
        ) {
            Text(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(0.dp, 10.dp),
                text = title,
                style = TextStyle(
                    fontSize = 20.sp,
                    textAlign = TextAlign.Center,
                    fontWeight = FontWeight.Bold
                )
            )
            TextField(
                state = textFieldState,
                modifier = Modifier
                    .padding(20.dp, 10.dp)
                    .height(40.dp),
                contentAlignment = Alignment.Center,
                hint = stringResource(R.string.dialog_input_string_input_hint),
                style = TextStyle(fontSize = 20.sp)
            )
            Divider(0.dp)
            Row(Modifier.height(45.dp)) {
                Box(
                    Modifier
                        .fillMaxHeight()
                        .weight(1f)
                        .clickable {
                            onDismissRequest()
                        }) {
                    Text(
                        modifier = Modifier.align(Alignment.Center),
                        text = stringResource(R.string.dialog_input_string_cancel),
                        style = TextStyle(
                            fontSize = 20.sp,
                            color = CHelperTheme.colors.mainColor,
                            textAlign = TextAlign.Center
                        )
                    )
                }
                DividerVertical(0.dp)
                Box(
                    Modifier
                        .fillMaxHeight()
                        .weight(1f)
                        .clickable {
                            onDismissRequest()
                            onConfirm()
                        }) {
                    Text(
                        modifier = Modifier.align(Alignment.Center),
                        text = stringResource(R.string.dialog_input_string_confirm),
                        style = TextStyle(
                            fontSize = 20.sp,
                            color = CHelperTheme.colors.mainColor,
                            textAlign = TextAlign.Center
                        )
                    )
                }
            }
        }
    }
}

@Preview
@Composable
fun InputStringDialogLightThemePreview() {
    CHelperTheme(
        theme = CHelperTheme.Theme.Light,
        backgroundBitmap = null
    ) {
        InputStringDialog(
            onDismissRequest = { },
            title = "请输入透明度",
            textFieldState = rememberTextFieldState(initialText = "100"),
            onConfirm = {}
        )
    }
}

@Preview
@Composable
fun InputStringDialogDarkThemePreview() {
    CHelperTheme(
        theme = CHelperTheme.Theme.Dark,
        backgroundBitmap = null
    ) {
        InputStringDialog(
            onDismissRequest = { },
            title = "请输入透明度",
            textFieldState = rememberTextFieldState(initialText = "100"),
            onConfirm = {}
        )
    }
}
